src/components/LocaleContextProvider.tsx

```diff
@@ -73,8 +80,35 @@ function LocaleContextProvider({children}: LocaleContextProviderProps) {
     const currentUserPersonalDetails = useCurrentUserPersonalDetails();
     const [areTranslationsLoading = true] = useOnyx(ONYXKEYS.ARE_TRANSLATIONS_LOADING, {initWithStoredValues: false, canBeMissing: true});
     const [countryCodeByIP = 1] = useOnyx(ONYXKEYS.COUNTRY_CODE, {canBeMissing: true});
+    const [nvpPreferredLocale, nvpPreferredLocaleMetadata] = useOnyx(ONYXKEYS.NVP_PREFERRED_LOCALE, {canBeMissing: true});
     const [currentLocale, setCurrentLocale] = useState<Locale | undefined>(() => IntlStore.getCurrentLocale());
 
+    const localeToApply = useMemo(() => {
+        if (isLoadingOnyxValue(nvpPreferredLocaleMetadata)) {
+            return undefined;
+        }
+        if (nvpPreferredLocale && isSupportedLocale(nvpPreferredLocale)) {
+            return nvpPreferredLocale;
+        }
+        return getDevicePreferredLocale();
+    }, [nvpPreferredLocale, nvpPreferredLocaleMetadata]);
```

Comment:
NAB - **Missing Type Guards**: `getDevicePreferredLocale()` might return a locale that's not supported, but there's no validation on the last return:
```suggestion
    const localeToApply = useMemo(() => {
        if (isLoadingOnyxValue(nvpPreferredLocaleMetadata)) {
            return undefined;
        }
        if (nvpPreferredLocale && isSupportedLocale(nvpPreferredLocale)) {
            return nvpPreferredLocale;
        }
        
        const deviceLocale = getDevicePreferredLocale();
        return isSupportedLocale(deviceLocale) ? deviceLocale : CONST.LOCALES.DEFAULT;
    }, [nvpPreferredLocale, nvpPreferredLocaleMetadata]);
```